src/snowflake/cli/api/utils/rendering.py

```diff
@@ -88,24 +96,122 @@ def jinja_render_from_file(
+        value = variables_data[key]
+        if not isinstance(value, str):
+            continue
+
+        try:  # try to evaluate the template given current state of know variables
+            new_value = env.from_string(value).render(context_data)
+            if string_includes_template(new_value):
+                unresolved_keys.append(key)
+            variables_data[key] = env.from_string(value).render(context_data)
+        except UndefinedError:
+            unresolved_keys.append(key)
+
+    return context_data
+
 
-    def __getattr__(self, item):
-        if item not in self._data_dict:
-            raise AttributeError(f"No attribute {item}")
-        return self._data_dict[item]
+def _check_variables_consistency(variables_data: DictWithEnvironFallback):
+    """
+    Checks consistency of provided dictionary by
+    1. checking reference cycles
+    2. checking for missing variables
+    """
+    # Variables that are not specified in env section
+    missing_variables: Set[str] = set()
+    # Variables that require other variables
+    variables_with_dependencies: Set[str] = set()
+
+    for key, value in variables_data.items():
+        # Templates are reserved only to string variables
+        if not isinstance(value, str):
+            continue
+
+        # This value requires
+        include_template, required_variables = _search_for_required_variables(value)
+
+        if key in required_variables:
+            raise ClickException(
+                f"Cycle detected in environment definition for `{key}` variable."
+            )
+
+        if required_variables or include_template:
+            variables_with_dependencies.add(key)
+
+        for variable in required_variables:
+            if variable not in variables_data:
+                missing_variables.add(variable)
+
+    # If there are unknown env variables then we raise an error
+    if missing_variables:
+        raise ClickException(
+            "The following variables are used in environment definition but are not defined: {}".format(
+                ", ".join(missing_variables)
+            )
+        )
+    return variables_with_dependencies
 
 
-def _add_project_context(data: Dict):
-    context_key = "ctx"
-    if context_key in data:
-        raise ValueError(f"{context_key} in user defined data")
-    context_data = {context_key: {"env": _AttrGetter(os.environ)}}
-    return {**data, **context_data}
+def _search_for_required_variables(variable_value: str):
+    """
+    Look for &{...} pattern in  variable value. Returns a list of env variables required
+    to expand this template.`
+    """
+    ctx_env_prefix = f"{_CONTEXT_KEY}.env."
+    found_variables = re.findall(r"(&{([\.\w ]+)})+", variable_value)
+    required_variables = []
+    for _, variable in found_variables:
+        var: str = variable.strip()
+        if var.startswith(ctx_env_prefix):
+            required_variables.append(var[len(ctx_env_prefix) :])
```

Comment:
I can error. This will cause that `env` variables cannot reference other parts of definition file. But this can be fine for now. Do we agree to raise error and remove support for cross-project references in `env` for now?

Reply:
yes, I agree to raise error for now. I will remove the error / tighten the check as part of my changes later.